src/main/java/com/epam/aidial/core/security/AccessTokenValidator.java

```diff
@@ -1,59 +1,154 @@
 
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
 
 public class AccessTokenValidator {
 
+    private static final long USER_INFO_EXP_PERIOD_MS = TimeUnit.MINUTES.toMillis(1);
+
     private final List<IdentityProvider> providers = new ArrayList<>();
 
-    public AccessTokenValidator(JsonObject idpConfig, Vertx vertx) {
+    private final ConcurrentMap<String, Future<UserInfoResult>> userInfoCache = new ConcurrentHashMap<>();
+
+    public AccessTokenValidator(JsonObject idpConfig, Vertx vertx, HttpClient client) {
         int size = idpConfig.size();
         if (size < 1) {
             throw new IllegalArgumentException("At least one identity provider is required");
         }
         for (String idpKey : idpConfig.fieldNames()) {
-            providers.add(new IdentityProvider(idpConfig.getJsonObject(idpKey), vertx, jwksUrl -> {
+            providers.add(new IdentityProvider(idpConfig.getJsonObject(idpKey), vertx, client, jwksUrl -> {
                 try {
                     return new UrlJwkProvider(new URL(jwksUrl));
                 } catch (MalformedURLException e) {
                     throw new IllegalArgumentException(e);
                 }
             }));
         }
+        vertx.setPeriodic(0, USER_INFO_EXP_PERIOD_MS, event -> evictExpiredUserInfo());
+    }
+
+    private void evictExpiredUserInfo() {
+        long currentTime = System.currentTimeMillis();
+        for (Map.Entry<String, Future<UserInfoResult>> entry : userInfoCache.entrySet()) {
+            UserInfoResult result = entry.getValue().result();
+            if (result != null && result.expirationTime() <= currentTime) {
+                userInfoCache.remove(entry.getKey());
+            }
+        }
     }
 
     public Future<ExtractedClaims> extractClaims(String authHeader) {
+        String accessToken = null;
         try {
             if (authHeader == null) {
                 return Future.succeededFuture();
             }
-            String encodedToken = Objects.requireNonNull(extractTokenFromHeader(authHeader), "Can't extract access token from header");
-            DecodedJWT jwt = IdentityProvider.decodeJwtToken(encodedToken);
+            accessToken = Objects.requireNonNull(extractTokenFromHeader(authHeader), "Can't extract access token from header");
             if (providers.size() == 1) {
-                return providers.get(0).extractClaims(jwt);
-            }
-            for (IdentityProvider idp : providers) {
-                if (idp.match(jwt)) {
-                    return idp.extractClaims(jwt);
-                }
+                IdentityProvider provider =  providers.get(0);
+                return extractClaims(accessToken, provider);
```

Comment:
I don't think that handling single IdP will benefit a lot. for now it only produce extra code